### PR TITLE
Implement FDBM in the GUI.

### LINKED ISSUE #137: Add FDBM support to the GUI
The full-density based model was implemented in #130. It's maybe a bit niche, but it could be added to the GUI as well. One thing that makes this simulation model different is that it needs three input files: the sample hartree potential, the sample electron density, and the tip electron density. This needs some rework in how the input files are loaded into the program. The two density overlap integral parameters, the scaling constant and the exponent, have to be added to the GUI as well (probably greyed out normally).

### PR DESCRIPTION
**Pull request by NikoOinonen**
Fixes #137 

Adds the option to use the FDBM model in the GUI with input boxes for the two parameters in the density overlap integral.

The file-open dialog is changed to be able to load several input files at once. If the input files contain a Hartree potential, a sample electron density, and a tip electron density, then the FDBM is automatically enabled.

Additionally, when using an LJ force field, it is now possible to load the tip electron density from a file for the electrostatic interaction, instead of using the multipole density. There is an option to load either a full electron density, from which the valence electrons are automatically subtracted, or you can load a delta density which is already neutral.

I will also update the wiki with instructions for opening the files afterwards.

### CODE REVIEW

**Comment by NikoOinonen** [comment]
There is still a problem with loading parameters when using the FDBM.

**Comment by NikoOinonen** [comment]
Ok, should work now.

**Comment by ProkopHapala** [comment]
Niko, that is great. 

Do we have some simple example for user to test it? I mean with Lennard-Jones based mehtod people need just .xyz so the barrier to give-it-a-try is low. But with FDBM one need large .xsf/.cube files form DFT, which is quite a barrier. Does the simple pyridine-FDBM example work with GUI. (It would be nice to prepare some example such that people can try it straitghtaway without much tinkering)

**Review by ProkopHapala — APPROVED** [review]
Great!

but pleas help me how to test it? I wanted to run it to see how the GUI looks now. But I probably still don't know how to use ppafm in developers mode. 

What I did (did not work) is:

1) clone to new instance of the repo to `git/TMP/ppafm` directory
```
cd git/TMP
git clone https://github.com/Probe-Particle/ppafm.git
```

2) `git checkout gui-fdbm`

3) install with localy with pip 
`pip install -e .[opencl]`

4) export python path
`export PYTHONPATH=/home/prokop/git/TMP/ppafm/`

5) got to `/git/TMP/ppafm/ppafm/cli/gui` and run the gui
`prokop@DesktopGTX3060:~/git/TMP/ppafm/ppafm/cli/gui$ python ppafm_gui.py`


But it crashed with the error:

```
prokop@DesktopGTX3060:~/git/TMP/ppafm/ppafm/cli/gui$ python ppafm_gui.py
 PACKAGE_PATH =  /home/prokop/git/TMP/ppafm/ppafm
 CPP_PATH     =  /home/prokop/git/TMP/ppafm/ppafm/cpp
Traceback (most recent call last):
  File "/home/prokop/git/TMP/ppafm/ppafm/cli/gui/ppafm_gui.py", line 18, in <module>
    import ppafm.GUIWidgets as guiw
  File "/home/prokop/git/TMP/ppafm/ppafm/GUIWidgets.py", line 9, in <module>
    from . import io
  File "/home/prokop/git/TMP/ppafm/ppafm/io.py", line 10, in <module>
    from .GridUtils import readNumsUpTo
  File "/home/prokop/git/TMP/ppafm/ppafm/GridUtils.py", line 11, in <module>
    lib = cpp_utils.get_cdll('GU')
  File "/home/prokop/git/TMP/ppafm/ppafm/cpp_utils.py", line 53, in get_cdll
    raise RuntimeError(f"Could not find compiled extension module in `{module_path}`. "
RuntimeError: Could not find compiled extension module in `/home/prokop/git/TMP/ppafm/ppafm/cpp/GridUtils_lib.so`. Either check pip installation or enable dynamic compilation by setting the environment variable PPAFM_RECOMPILE=1
```

I guess it is the same issue as (#135 ), maybe it is not really some issue it is perhaps I don't know how to run the code now.

File: ppafm/cli/gui/ppafm_gui.py
Comment (NikoOinonen) [reply ppafm/cli/gui/ppafm_gui.py:90]: Yes, the device cannot be currently set from within the GUI. There is an existing issue #82 for this.

**Comment by NikoOinonen** [comment]
@ProkopHapala BTW, you don't have to be so quick to approve the PR if you think there are still things to be fixed or improved. I don't think this feature is in so much in hurry to be merged to the main branch.

**Comment by ProkopHapala** [comment]
> @ProkopHapala BTW, you don't have to be so quick to approve the PR if you think there are still things to be fixed or improved. I don't think this feature is in so much in hurry to be merged to the main branch.

OK. But I generally think it is better that fetures which are not controversial or obviously wrong are in the main branch as fast ans possible. I think this Pull-requst pollicy is good mainly to let other people know what is going on, and open some disscussion, but I don't think it is good to block people, ... especially you when you are so productive :)

**Comment by ProkopHapala** [comment]
> The main input is either the geometry from xyz, .in, POSCAR, or its the Hartree potential from .xsf or .cube. This is the only required input. If the other fields are empty, then yes it's just LJ + electrostatics either from point charges or the Hartree potential convolution.

Generally, what I wanted to say is *explicit is better than implicit*, so maybe it would be better to really have explicity switch which mode you use (LJ+point charge, LJ+electrostatic convolution, FDBM etc.), and based on that open different dialog menu, or at least shade-out some filelds which are not relevant for that mode of operation. This will let user to think more clearly about what it is doing and what input he needs.

**Comment by NikoOinonen** [comment]
> OK. But I generally think it is better that fetures which are not controversial or obviously wrong are in the main branch as fast ans possible. I think this Pull-requst pollicy is good mainly to let other people know what is going on, and open some disscussion, but I don't think it is good to block people, ... especially you when you are so productive :)

Actually, I kind of think the opposite. The main branch is supposed to be the version that everyone more-or-less agrees is good, so it's better to take some time to think things through. If the behaviour of the code is changing back and forth often in a way that can affect other peoples work, then it just gets annoying for anyone who is not in on the discussion for the current changes. And I think it's better to have all the discussion related to a specific feature in one place, if possible. Once a PR is merged, it's final, and another one needs to be opened if further changes are needed. If I really needed someone else to have this feature right now, I would just tell them to use the separate branch until things settle.

> Generally, what I wanted to say is explicit is better than implicit, so maybe it would be better to really have explicity switch which mode you use (LJ+point charge, LJ+electrostatic convolution, FDBM etc.), and based on that open different dialog menu, or at least shade-out some filelds which are not relevant for that mode of operation. This will let user to think more clearly about what it is doing and what input he needs.

Yes, I like this idea. We could have a [radio button](https://en.wikipedia.org/wiki/Radio_button) or tabs to choose the simulation type, and change the available boxes based on the chosen option. This is exactly the kind of feedback that I want to get before I push it for everyone to see. I'll try to implement this soon.

**Comment by NikoOinonen** [comment]
I updated the file open dialog to use tabs for the different force field models. It looks like this:
image
Now it only shows the input boxes relevant to the specific simulation model and provides an explanation, so hopefully it should be more clear now.

@ProkopHapala Could you check that it's okay now?

**Comment by ProkopHapala** [comment]
> @ProkopHapala Could you check that it's okay now?

Very nice! It works and it is much more clear now. 

Two things: 
1) I was thinking that in FDBM the Hartree potential is also optional because you calculate it by solving poisson equation? 
2) Why there is no checkbox like "subtract neutral atom density" in FDBM?

**Comment by NikoOinonen** [comment]
> I was thinking that in FDBM the Hartree potential is also optional because you calculate it by solving poisson equation?

We don't have that option yet. It's an open issue #150.

> Why there is no checkbox like "subtract neutral atom density" in FDBM?

Because you always need the tip electron density for the Pauli integral, so it always also does the subtraction if the user doesn't supply a delta density.

**Pull request merged**
